bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -161,11 +162,15 @@ void KeyExchangeManager::onPublishClientsKeys(const std::string& keys, std::opti
       LOG_FATAL(KEY_EX_LOG, "Initial Published Clients keys and replica client keys do not match");
       ConcordAssertEQ(keys, *bootstrap_keys);
     }
-    if (impl::KeyExchangeManager::instance().clientKeysPublished()) save = false;
-  }
-  if (save) {
-    impl::KeyExchangeManager::instance().saveClientsPublicKeys(keys);
+    if (clientKeysPublished()) save = false;
   }
+  if (save) saveClientsPublicKeys(keys);
+}
+
+void KeyExchangeManager::onClientPublicKeyExchange(const std::string& key, std::uint16_t clientId) {
```

Comment:
On client key exchange we need to set ```ClientKeyStore::published_``` to false and I think we need to call ```sendInitialClientsKeys``` with ```SigManager::instance()->getClientsPublicKeys()``` .
This should publish the new key set